stock_valuation_specific_identification/models/product.py

```diff
@@ -0,0 +1,154 @@
+
+    def _run_fifo(self, quantity, company):
+        if not self.specific_ident_cost or self.product_tmpl_id.tracking == "none":
+            return super(ProductProduct, self)._run_fifo(quantity, company)
+
+        self.ensure_one()
+        move = self.env.context.get("move", False)
+        if move:
+            rounding = move.product_uom.rounding
+            valued_move_lines = move.move_line_ids.filtered(
+                lambda ml: ml.location_id._should_be_valued()
+                and not ml.location_dest_id._should_be_valued()
+                and not ml.owner_id
+            )
+            candidates = (
+                self.env["stock.valuation.layer"]
+                .sudo()
+                .search(
+                    [
+                        ("product_id", "=", self.id),
+                        ("remaining_qty", ">", 0),
+                        ("company_id", "=", company.id),
+                    ]
+                )
+            )
+            # extra lots
+            lots = valued_move_lines.mapped("lot_id")
+            new_standard_price = 0
+            tmp_value = 0  # to accumulate the value taken on the candidates
+            qty_to_take_on_lots = {x: 0.0 for x in lots}
+
+            for valued_move_line in valued_move_lines:
+                lot = valued_move_line.lot_id
+                qty_to_take_on_candidates = (
+                    valued_move_line.product_uom_id._compute_quantity(
+                        valued_move_line.qty_done, move.product_id.uom_id
+                    )
+                )
+                qty_to_take_on_lots[lot] += qty_to_take_on_candidates
+                for candidate in candidates:
+                    if (
+                        float_compare(
+                            candidate.remaining_qty,
+                            sum(
+                                candidate.stock_move_id.move_line_ids.mapped(
+                                    "remaining_qty"
+                                )
+                            ),
+                            precision_rounding=rounding,
+                        )
+                        != 0
+                    ):
+                        raise UserError(
+                            _(
+                                "Line remaining quantity does not match "
+                                "move remaining quantity for move %s"
+                                % candidate.stock_move_id.name
+                            )
+                        )
+                    for (
+                        candidate_line
+                    ) in candidate.stock_move_id.move_line_ids.filtered(
+                        lambda x: x.lot_id == lot and x.remaining_qty > 0.0
+                    ):
+                        qty_taken_on_candidate = min(
+                            qty_to_take_on_candidates, candidate_line.remaining_qty
+                        )
```

Comment:
I don't understand this part. `candidate_line` is an instance of `stock move line`, isn't? I don't think `stock move line` record has `remaining_qty` field. Am I missing something?